subsys/bluetooth/audio/has_client.c

```diff
@@ -0,0 +1,770 @@
+
+		call_has_preset_cb(&client->has, gu->idx, gu->properties, gu->name, buf->len);
+
+	} else {
+		uint8_t index;
+
+		if (buf->len < sizeof(index)) {
+			/* discard the notification */
+			return;
+		}
+
+		index = net_buf_simple_pull_u8(buf);
+
+		if (pc->change_id == BT_HAS_CHANGE_ID_PRESET_DELETED) {
+			has_cb->preset_deleted(&client->has, index);
+		} else if (pc->change_id == BT_HAS_CHANGE_ID_PRESET_AVAILABLE) {
+			has_cb->preset_availibility(&client->has, index, true);
+		} else if (pc->change_id == BT_HAS_CHANGE_ID_PRESET_UNAVAILABLE) {
+			has_cb->preset_availibility(&client->has, index, false);
+		}
+	}
+}
+
+static uint8_t control_point_rx(struct bt_conn *conn, struct bt_gatt_subscribe_params *params,
+				const void *data, uint16_t len)
+{
+	struct bt_has_client *client = client_by_conn(conn);
+	const struct bt_has_cp_hdr *hdr;
+	struct net_buf_simple buf;
+
+	if (len < sizeof(*hdr)) {
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	net_buf_simple_init_with_data(&buf, (void *)data, len);
+
+	hdr = net_buf_simple_pull_mem(&buf, sizeof(*hdr));
+
+	BT_DBG("conn %p buf %p len %u op %s (0x%02x)", conn, data, len, bt_has_op_str(hdr->op),
+	       hdr->op);
+
+	switch (hdr->op) {
+	case BT_HAS_OP_READ_PRESET_RSP:
+		cp_read_preset_rsp(client, &buf);
+		break;
+	case BT_HAS_OP_PRESET_CHANGED:
+		cp_preset_changed_ind(client, &buf);
+		break;
+	}
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static void cp_subscribe_cb(struct bt_conn *conn, uint8_t err, struct bt_gatt_write_params *write)
+{
+	struct bt_has_client *client = client_by_conn(conn);
+
+	BT_DBG("conn %p err 0x%02x", client->conn, err);
+
+	if (err || disc_active_preset_idx(client) < 0) {
+		disc_complete(client, false);
+	}
+}
+
+static int cp_subscribe(struct bt_has_client *client, uint16_t handle)
+{
+	BT_DBG("conn %p handle 0x%04x", client->conn, handle);
+
+	client->cp_subscription.notify = control_point_rx;
+	client->cp_subscription.write = cp_subscribe_cb;
+	client->cp_subscription.value_handle = handle;
+	client->cp_subscription.ccc_handle = 0x0000;
+	client->cp_subscription.end_handle = 0xffff;
+	client->cp_subscription.disc_params = &client->params.discover;
+	client->cp_subscription.value = BT_GATT_CCC_INDICATE;
+	atomic_set_bit(client->cp_subscription.flags, BT_GATT_SUBSCRIBE_FLAG_VOLATILE);
+
+	return bt_gatt_subscribe(client->conn, &client->cp_subscription);
+}
+
+static uint8_t cp_discover_cb(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			      struct bt_gatt_discover_params *discover)
+{
+	struct bt_has_client *client = client_by_conn(conn);
+	struct bt_gatt_chrc *chrc;
+
+	BT_DBG("conn %p attr %p", client->conn, attr);
+
+	if (!attr) {
+		BT_INFO("HAS Control Point not found");
+		disc_complete(client, true);
+		return BT_GATT_ITER_STOP;
+	}
+
+	chrc = attr->user_data;
+
+	if (cp_subscribe(client, chrc->value_handle) < 0) {
+		disc_complete(client, false);
+	}
+
+	return BT_GATT_ITER_STOP;
+}
+
+static int cp_discover(struct bt_has_client *client)
```

Comment:
Should this be `disc_cp`, for consistency with the discovery functions for the two other characteristics?